hazelcast/src/hazelcast/client/connection/ClientConnectionManagerImpl.cpp

```diff
@@ -0,0 +1,842 @@
+            }
+
+            void
+            ClientConnectionManagerImpl::fireConnectionEvent(
+                    const hazelcast::client::LifecycleEvent::LifeCycleState &state) {
+                spi::LifecycleService &lifecycleService = client.getLifecycleService();
+                lifecycleService.fireLifecycleEvent(state);
+            }
+
+            boost::shared_ptr<util::Future<bool> > ClientConnectionManagerImpl::connectToClusterAsync() {
+                boost::shared_ptr<util::Callable<bool> > task(new ConnectToClusterTask(*this));
+                return clusterConnectionExecutor->submit<bool>(task);
+            }
+
+            void ClientConnectionManagerImpl::connectToClusterInternal() {
+                int attempt = 0;
+                std::set<Address> triedAddresses;
+
+                while (attempt < connectionAttemptLimit) {
+                    attempt++;
+                    int64_t nextTry = util::currentTimeMillis() + connectionAttemptPeriod;
+
+                    std::set<Address> addresses = getPossibleMemberAddresses();
+                    BOOST_FOREACH (const Address &address, addresses) {
+                                    if (!client.getLifecycleService().isRunning()) {
+                                        throw exception::IllegalStateException(
+                                                "ConnectionManager::connectToClusterInternal",
+                                                "Giving up on retrying to connect to cluster since client is shutdown.");
+                                    }
+                                    triedAddresses.insert(address);
+                                    if (connectAsOwner(address).get() != NULL) {
+                                        return;
+                                    }
+                                }
+
+                    // If the address providers load no addresses (which seems to be possible), then the above loop is not entered
+                    // and the lifecycle check is missing, hence we need to repeat the same check at this point.
+                    if (!client.getLifecycleService().isRunning()) {
+                        throw exception::IllegalStateException("Client is being shutdown.");
+                    }
+
+                    if (attempt < connectionAttemptLimit) {
+                        const int64_t remainingTime = nextTry - util::currentTimeMillis();
+                        logger.warning() << "Unable to get alive cluster connection, try in "
+                                         << (remainingTime > 0 ? remainingTime : 0) << " ms later, attempt " << attempt
+                                         << " of " << connectionAttemptLimit << ".";
+
+                        if (remainingTime > 0) {
+                            // TODO: change with interruptible sleep
+                            util::sleepmillis(remainingTime);
```

Comment:
Thread reference can be passed to here to achieve that. 
I think all Callable/Runnnable's should carry reference to their thread so that we can call interruptible sleep.

Reply:
I added a static method static void AbstractThread::sleep(int64_t timeInMilliseconds);